File: lib/eip8128_signer.py
"""
EIP-8128 HTTP Message Signature Signer for KarmaCadabra Agents

Signs HTTP requests per ERC-8128 (Signed HTTP Requests with Ethereum)
so that the Execution Market server can authenticate agents by their
wallet address without API keys.

Produces three HTTP headers:
  - Content-Digest   (POST/PUT/PATCH with body only)
  - Signature-Input  (RFC 9421 structured field)
  - Signature        (65-byte secp256k1 signature, base64 in structured field)

The server verifier lives at mcp_server/integrations/erc8128/verifier.py.
This signer is its exact complement -- the signature base construction,
content digest format, and EIP-191 signing all mirror the verifier.

Reference:
  - ERC-8128: https://eip.tools/eip/8128
  - RFC 9421: https://www.rfc-editor.org/rfc/rfc9421
  - ERC-191:  https://eips.ethereum.org/EIPS/eip-191
"""

import base64
import hashlib
import logging
import secrets
import time
from urllib.parse import urlparse

import requests
from eth_account import Account
from eth_account.messages import encode_defunct

logger = logging.getLogger("kk.eip8128_signer")

# Methods that carry a request body per HTTP semantics
_BODY_METHODS = {"POST", "PUT", "PATCH"}

# Default validity window in seconds (matches TypeScript signer and server max)
_DEFAULT_VALIDITY_SEC = 300


class EIP8128Signer:
    """Sign HTTP requests with EIP-8128 wallet-based authentication.

    The signer derives the wallet address from the private key and
    constructs RFC 9421-compliant signature base strings signed with
    EIP-191 personal_sign.

    Args:
        private_key: Hex-encoded Ethereum private key (with or without 0x prefix).
        chain_id: EVM chain ID for the keyid (default: 8453 = Base Mainnet).
        api_base: Base URL of the Execution Market API. Used for nonce fetching.
        validity_sec: Signature validity window in seconds (default: 60, max 300).
    """

    def __init__(
        self,
        private_key: str,
        chain_id: int = 8453,
        api_base: str = "https://api.execution.market",
        validity_sec: int = _DEFAULT_VALIDITY_SEC,
    ):
        self._account = Account.from_key(private_key)
        self._chain_id = chain_id
        self._api_base = api_base.rstrip("/")
        self._validity_sec = min(validity_sec, 300)  # Server max is 300s

    @property
    def address(self) -> str:
        """Wallet address (checksummed)."""
        return self._account.address

    @property
    def address_lower(self) -> str:
        """Wallet address (lowercase, as used in keyid)."""
        return self._account.address.lower()

    @property
    def keyid(self) -> str:
        """ERC-8128 keyid string: ``erc8128:<chainId>:<0xaddress>``."""
        return f"erc8128:{self._chain_id}:{self.address_lower}"

    def sign_request(
        self,
        method: str,
        url: str,
        body: str = "",
    ) -> dict[str, str]:
        """Sign an HTTP request and return the authentication headers.

        Args:
            method: HTTP method (GET, POST, PUT, PATCH, DELETE, etc.).
            url: Full request URL (e.g. ``https://api.execution.market/api/v1/tasks``).
            body: Request body string. Only relevant for POST/PUT/PATCH.

        Returns:
            Dictionary of headers to merge into the outgoing request:
            ``content-digest``, ``signature-input``, ``signature``.
            For GET/DELETE/HEAD (no body), ``content-digest`` is omitted.
        """
        method_upper = method.upper()
        parsed = urlparse(url)
        authority = parsed.netloc  # e.g. "api.execution.market"
        path = parsed.path or "/"

        # --- 1. Content-Digest (only for methods with a body) ----------------
        has_body = method_upper in _BODY_METHODS and len(body) > 0
        content_digest_value: str | None = None
        if has_body:
            content_digest_value = self._compute_content_digest(body)

        # --- 2. Covered components -------------------------------------------
        covered: list[str] = ["@method", "@authority", "@path"]
        if has_body:
            covered.append("content-digest")

        # --- 3. Signature parameters -----------------------------------------
        nonce, server_ttl = self._fetch_nonce()
        now = int(time.time())
        created = now
        validity = server_ttl if server_ttl else self._validity_sec
        expires = now + min(validity, 300)

        params: dict[str, object] = {
            "created": created,
            "expires": expires,
            "nonce": nonce,
            "keyid": self.keyid,
        }

        sig_params_str = self._build_signature_params(covered, params)

        # --- 4. Signature base (RFC 9421) ------------------------------------
        sig_base = self._build_signature_base(
            method=method_upper,
            authority=authority,
            path=path,
            content_digest_value=content_digest_value,
            covered=covered,
            sig_params_str=sig_params_str,
        )

        # --- 5. EIP-191 sign ------------------------------------------------
        sig_bytes = self._eip191_sign(sig_base)
        sig_b64 = base64.b64encode(sig_bytes).decode("ascii")

        # --- 6. Build response headers ---------------------------------------
        headers: dict[str, str] = {}
        if content_digest_value is not None:
            headers["content-digest"] = content_digest_value
        headers["signature-input"] = f"eth={sig_params_str}"
        headers["signature"] = f"eth=:{sig_b64}:"

        return headers

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _fetch_nonce(self) -> tuple[str, int | None]:
        """Fetch a fresh nonce from the server.

        Calls ``GET /api/v1/auth/erc8128/nonce`` on the Execution Market API.
        If the endpoint is unavailable (404, network error, etc.),
        falls back to a locally generated nonce.

        Returns:
            Tuple of (nonce_string, ttl_seconds). ttl_seconds is None when
            the server did not provide a TTL or the endpoint was unavailable.
        """
        nonce_url = f"{self._api_base}/api/v1/auth/erc8128/nonce"
        try:
            resp = requests.get(nonce_url, timeout=5)
            if resp.status_code == 200:
                data = resp.json()
                nonce = data.get("nonce")
                if nonce:
                    ttl = data.get("ttl_seconds")
                    return str(nonce), ttl if isinstance(ttl, int) else None
        except Exception as exc:
            logger.debug("Nonce endpoint unavailable (%s), using local nonce", exc)

        # Fallback: generate a locally unique nonce
        return secrets.token_urlsafe(24), None

    def _build_signature_base(
        self,
        method: str,
        authority: str,
        path: str,
        content_digest_value: str | None,
        covered: list[str],
        sig_params_str: str,
    ) -> str:
        """Build the RFC 9421 signature base string.

        This mirrors ``_build_signature_base()`` in the server verifier
        (``mcp_server/integrations/erc8128/verifier.py``).

        Format (each component on its own line, no trailing newline)::

            "@method": POST
            "@authority": api.execution.market
            "@path": /api/v1/tasks
            "content-digest": sha-256=:base64hash:
            "@signature-params": ("@method" ...);created=...;...

        Args:
            method: HTTP method (uppercase).
            authority: Request authority (host[:port]).
            path: Request path.
            content_digest_value: Full Content-Digest header value, or None.
            covered: List of covered component identifiers.
            sig_params_str: Pre-built @signature-params value string.

        Returns:
            The signature base string ready for EIP-191 signing.
        """
        lines: list[str] = []

        for component in covered:
            value = self._resolve_component(
                component, method, authority, path, content_digest_value
            )
            lines.append(f'"{component}": {value}')

        lines.append(f'"@signature-params": {sig_params_str}')

        return "\n".join(lines)

    @staticmethod
    def _resolve_component(
        component: str,
        method: str,
        authority: str,
        path: str,
        content_digest_value: str | None,
    ) -> str:
        """Resolve a single covered component's value.

        This mirrors ``_resolve_component()`` in the server verifier.
        """
        if component == "@method":
            return method.upper()
        if component == "@authority":
            return authority
        if component == "@path":
            return path
        if component == "content-digest":
            return content_digest_value or ""
        # Unknown component -- return empty string for forward compatibility
        return ""

    @staticmethod
    def _build_signature_params(covered: list[str], params: dict[str, object]) -> str:
        """Build the @signature-params structured field value.

        This mirrors ``_build_signature_params()`` in the server verifier.
        Parameter order: created, expires, nonce, keyid (then any extras
        sorted alphabetically).

        Returns:
            String like ``("@method" "@authority" "@path");created=N;expires=N;nonce="v";keyid="erc8128:..."``
        """
        comp_str = " ".join(f'"{c}"' for c in covered)
        parts: list[str] = [f"({comp_str})"]

        ordered_keys = ["created", "expires", "nonce", "keyid"]
        for key in ordered_keys:
            if key in params:
                parts.append(_format_param(key, params[key]))
        for key in sorted(params.keys()):
            if key not in ordered_keys:
                parts.append(_format_param(key, params[key]))

        return ";".join(parts)

    @staticmethod
    def _compute_content_digest(body: str) -> str:
        """Compute the Content-Digest header value for a request body.

        Format: ``sha-256=:base64(sha256(body)):``

        This mirrors ``_verify_content_digest()`` in the server verifier --
        the verifier expects the body hashed as raw bytes and the digest
        encoded in ``sha-256=:base64:`` format (RFC 9530).

        Args:
            body: The request body as a string.

        Returns:
            The full Content-Digest header value.
        """
        body_bytes = body.encode("utf-8")
        digest = hashlib.sha256(body_bytes).digest()
        b64 = base64.b64encode(digest).decode("ascii")
        return f"sha-256=:{b64}:"

    def _eip191_sign(self, message: str) -> bytes:
        """Sign a message with EIP-191 personal_sign.

        Per ERC-8128 Section 3.4.3::

            H = keccak256("\\x19Ethereum Signed Message:\\n" + len(M) + M)

        The server verifier uses ``encode_defunct(text=message)`` from
        eth_account, which produces the same EIP-191 prefix. We use the
        same function here for exact parity.

        Args:
            message: The signature base string to sign.

        Returns:
            65-byte signature (r[32] + s[32] + v[1]).
        """
        msg = encode_defunct(text=message)
        signed = self._account.sign_message(msg)
        return bytes(signed.signature)


def _format_param(key: str, value: object) -> str:
    """Format a single parameter for @signature-params.

    Integers are bare, everything else is quoted.
    Mirrors ``_format_param()`` in the server verifier.
    """
    if isinstance(value, int):
        return f"{key}={value}"
    return f'{key}="{value}"'


# ======================================================================
# Round-trip test
# ======================================================================

if __name__ == "__main__":
    import sys

    # Use a well-known test private key (Hardhat account #0)
    TEST_PK = "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80"

    signer = EIP8128Signer(
        TEST_PK,
        chain_id=8453,
        api_base="https://api.execution.market",
    )

    print(f"Signer address: {signer.address}")
    print(f"Signer keyid:   {signer.keyid}")
    print()

    # --- Test 1: POST with body ---
    print("=== POST /api/v1/tasks (with body) ===")
    post_headers = signer.sign_request(
        "POST",
        "https://api.execution.market/api/v1/tasks",
        '{"title":"test","bounty_usd":0.10}',
    )
    for k, v in post_headers.items():
        print(f"  {k}: {v}")
    print()

    # --- Test 2: GET without body ---
    print("=== GET /api/v1/tasks/available (no body) ===")
    get_headers = signer.sign_request(
        "GET",
        "https://api.execution.market/api/v1/tasks/available",
    )
    for k, v in get_headers.items():
        print(f"  {k}: {v}")
    print()

    # --- Test 3: Verify round-trip with the server verifier ---
    print("=== Round-trip verification against server verifier ===")
    try:
        sys.path.insert(
            0,
            str(
                __import__("pathlib")
                .Path(__file__)
                .resolve()
                .parent.parent.parent.parent
                / "mcp_server"
            ),
        )
        import asyncio
        from integrations.erc8128.verifier import verify_erc8128_request, VerifyPolicy
        from integrations.erc8128.nonce_store import InMemoryNonceStore

        class _MockURL:
            def __init__(self, path: str, query: str, netloc: str):
                self.path = path
                self.query = query
                self.netloc = netloc

        class _MockRequest:
            def __init__(
                self, method: str, url: _MockURL, headers: dict, body_bytes: bytes
            ):
                self.method = method
                self.url = url
                self.headers = headers
                self._body = body_bytes

            async def body(self):
                return self._body

        async def _verify():
            store = InMemoryNonceStore()

            # POST with body
            body_str = '{"title":"roundtrip","bounty_usd":0.05}'
            headers = signer.sign_request(
                "POST",
                "https://api.execution.market/api/v1/tasks",
                body_str,
            )
            all_headers = {**headers}
            req = _MockRequest(
                method="POST",
                url=_MockURL(
                    path="/api/v1/tasks",
                    query="",
                    netloc="api.execution.market",
                ),
                headers=all_headers,
                body_bytes=body_str.encode("utf-8"),
            )
            result = await verify_erc8128_request(
                req,
                nonce_store=store,
                policy=VerifyPolicy(allow_replayable=False),
            )
            if result.ok:
                print(
                    f"  POST verify: PASS (address={result.address}, chain={result.chain_id})"
                )
            else:
                print(f"  POST verify: FAIL ({result.reason})")
                sys.exit(1)

            # GET without body
            get_hdrs = signer.sign_request(
                "GET",
                "https://api.execution.market/api/v1/tasks/available",
            )
            req2 = _MockRequest(
                method="GET",
                url=_MockURL(
                    path="/api/v1/tasks/available",
                    query="",
                    netloc="api.execution.market",
                ),
                headers=get_hdrs,
                body_bytes=b"",
            )
            result2 = await verify_erc8128_request(
                req2,
                nonce_store=store,
                policy=VerifyPolicy(allow_replayable=False),
            )
            if result2.ok:
                print(
                    f"  GET  verify: PASS (address={result2.address}, chain={result2.chain_id})"
                )
            else:
                print(f"  GET  verify: FAIL ({result2.reason})")
                sys.exit(1)

            print()
            print("All round-trip verifications passed.")

        asyncio.run(_verify())

    except ImportError as exc:
        print(f"  Skipping round-trip (import error: {exc})")
        print("  Run from project root with mcp_server deps installed to verify.")

File: scripts/em-integration/lib/eip8128-signer.ts
/**
 * Karma Kadabra V2 — EIP-8128 HTTP Request Signer
 *
 * Signs HTTP requests per ERC-8128 (Signed HTTP Requests with Ethereum),
 * producing 3 headers: Content-Digest, Signature, and Signature-Input.
 *
 * Compatible with the server-side verifier at:
 *   mcp_server/integrations/erc8128/verifier.py
 *
 * References:
 *   - ERC-8128: https://eip.tools/eip/8128
 *   - RFC 9421: https://www.rfc-editor.org/rfc/rfc9421 (HTTP Message Signatures)
 *   - RFC 9530: https://www.rfc-editor.org/rfc/rfc9530 (Content-Digest)
 *   - EIP-191:  https://eips.ethereum.org/EIPS/eip-191 (personal_sign)
 *
 * Usage:
 *   import { EIP8128Signer, createSignedFetch } from "./eip8128-signer";
 *
 *   const signer = new EIP8128Signer("0xprivatekey", 8453);
 *   const headers = await signer.signRequest(
 *     "POST",
 *     "https://api.execution.market/api/v1/tasks",
 *     JSON.stringify({ title: "test" }),
 *   );
 *   // headers = { "content-digest": "...", "signature": "...", "signature-input": "..." }
 *
 *   // Or use the fetch wrapper:
 *   const signedFetch = createSignedFetch("0xprivatekey", 8453);
 *   const res = await signedFetch("https://api.execution.market/api/v1/tasks", {
 *     method: "POST",
 *     body: JSON.stringify({ title: "test" }),
 *   });
 */

import { createHash, randomBytes } from "crypto";
import {
  type Hex,
  type Address,
} from "viem";
import { type PrivateKeyAccount, privateKeyToAccount } from "viem/accounts";

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/** Default signature validity window in seconds */
const DEFAULT_VALIDITY_SEC = 300;

/** Signature label used in Signature / Signature-Input headers */
const SIGNATURE_LABEL = "eth";

/** API endpoint for server-issued nonces (canonical ERC-8128 path) */
const NONCE_ENDPOINT = "/api/v1/auth/erc8128/nonce";

/** HTTP methods that carry a request body */
const BODY_METHODS = new Set(["POST", "PUT", "PATCH"]);

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** Parsed URL components needed for signature base construction */
interface UrlComponents {
  /** Host without protocol (e.g. "api.execution.market") */
  authority: string;
  /** Path without query (e.g. "/api/v1/tasks") */
  path: string;
  /** Full origin for nonce requests (e.g. "https://api.execution.market") */
  origin: string;
}

// ---------------------------------------------------------------------------
// EIP8128Signer
// ---------------------------------------------------------------------------

/**
 * Signs HTTP requests per ERC-8128 using an Ethereum private key.
 *
 * Produces RFC 9421-compliant HTTP Message Signature headers that the
 * server-side verifier (`mcp_server/integrations/erc8128/verifier.py`)
 * can validate via EIP-191 ecrecover.
 *
 * @example
 * ```ts
 * const signer = new EIP8128Signer("0xabc123...", 8453);
 * const headers = await signer.signRequest(
 *   "POST",
 *   "https://api.execution.market/api/v1/tasks",
 *   '{"title":"Buy coffee"}',
 * );
 * ```
 */
export class EIP8128Signer {
  private readonly account: PrivateKeyAccount;
  private readonly chainId: number;
  private readonly address: Address;

  /**
   * @param privateKey - Hex-encoded private key (with or without "0x" prefix)
   * @param chainId    - EVM chain ID (default: 8453 = Base Mainnet)
   */
  constructor(privateKey: string, chainId: number = 8453) {
    const key = (privateKey.startsWith("0x") ? privateKey : `0x${privateKey}`) as Hex;
    this.account = privateKeyToAccount(key);
    this.chainId = chainId;
    this.address = this.account.address;
  }

  /** The signer's Ethereum address (checksummed) */
  get signerAddress(): Address {
    return this.address;
  }

  /**
   * Sign an HTTP request and return the ERC-8128 headers.
   *
   * For POST/PUT/PATCH with a body, returns 3 headers:
   *   - `content-digest`  — SHA-256 of the body (RFC 9530)
   *   - `signature`       — Base64-encoded 65-byte secp256k1 signature
   *   - `signature-input` — RFC 9421 structured field with covered components
   *
   * For GET/DELETE (no body), returns 2 headers (no content-digest).
   *
   * @param method - HTTP method (GET, POST, PUT, PATCH, DELETE)
   * @param url    - Full URL (e.g. "https://api.execution.market/api/v1/tasks")
   * @param body   - Request body string (required for POST/PUT/PATCH)
   * @returns Record of lowercase header names to their values
   */
  async signRequest(
    method: string,
    url: string,
    body?: string,
  ): Promise<Record<string, string>> {
    const upperMethod = method.toUpperCase();
    const { authority, path, origin } = EIP8128Signer.parseUrl(url);
    const hasBody = BODY_METHODS.has(upperMethod) && body != null && body.length > 0;

    // 1. Build Content-Digest if we have a body
    let contentDigest: string | null = null;
    if (hasBody) {
      contentDigest = computeContentDigest(body!);
    }

    // 2. Fetch nonce from server (with local fallback)
    const { nonce, ttlSeconds } = await this.fetchNonce(origin);

    // 3. Compute timestamps
    const now = Math.floor(Date.now() / 1000);
    const created = now;
    const expires = now + ttlSeconds;

    // 4. Build the keyid
    const keyid = `erc8128:${this.chainId}:${this.address.toLowerCase()}`;

    // 5. Determine covered components
    //    For body-bearing methods: include content-digest
    //    For GET/DELETE: omit content-digest
    const coveredComponents = hasBody
      ? ["@method", "@authority", "@path", "content-digest"]
      : ["@method", "@authority", "@path"];

    // 6. Build signature params string (matches verifier's _build_signature_params)
    //    Order: created, expires, nonce, keyid (then any extras sorted)
    const sigParams = buildSignatureParams(coveredComponents, {
      created,
      expires,
      nonce,
      keyid,
    });

    // 7. Build signature base (RFC 9421 Section 2.5)
    const sigBase = buildSignatureBase(
      coveredComponents,
      {
        "@method": upperMethod,
        "@authority": authority,
        "@path": path,
        "content-digest": contentDigest ?? "",
      },
      sigParams,
    );

    // 8. Sign with EIP-191 personal_sign
    const signatureBytes = await this.eip191Sign(sigBase);

    // 9. Encode signature as base64
    const signatureB64 = Buffer.from(signatureBytes).toString("base64");

    // 10. Assemble headers
    const headers: Record<string, string> = {};

    if (contentDigest) {
      headers["content-digest"] = contentDigest;
    }

    // Signature header: eth=:base64:
    headers["signature"] = `${SIGNATURE_LABEL}=:${signatureB64}:`;

    // Signature-Input header: eth=<sig-params>
    headers["signature-input"] = `${SIGNATURE_LABEL}=${sigParams}`;

    return headers;
  }

  /**
   * Fetch a nonce from the server's nonce endpoint.
   * Falls back to a locally-generated random nonce if the endpoint is unavailable.
   *
   * @param origin - Server origin (e.g. "https://api.execution.market")
   * @returns Object with nonce string and TTL in seconds
   */
  async fetchNonce(origin: string): Promise<{ nonce: string; ttlSeconds: number }> {
    const nonceUrl = `${origin}${NONCE_ENDPOINT}`;

    try {
      const response = await fetch(nonceUrl, {
        method: "GET",
        headers: { Accept: "application/json" },
        signal: AbortSignal.timeout(5000),
      });

      if (response.ok) {
        const data = (await response.json()) as { nonce?: string; ttl_seconds?: number };
        if (data.nonce) {
          return {
            nonce: data.nonce,
            ttlSeconds: data.ttl_seconds ?? DEFAULT_VALIDITY_SEC,
          };
        }
      }

      // Non-200 or missing nonce field — fall through to local generation
    } catch {
      // Network error, timeout, or 404 — fall through to local generation
    }

    // Fallback: generate a local nonce (16 random bytes, hex-encoded)
    return {
      nonce: randomBytes(16).toString("hex"),
      ttlSeconds: DEFAULT_VALIDITY_SEC,
    };
  }

  /**
   * Parse a full URL into authority and path components.
   *
   * @param url - Full URL string (e.g. "https://api.execution.market/api/v1/tasks")
   * @returns Object with authority, path, and origin
   *
   * @example
   * ```ts
   * EIP8128Signer.parseUrl("https://api.execution.market/api/v1/tasks")
   * // { authority: "api.execution.market", path: "/api/v1/tasks", origin: "https://api.execution.market" }
   * ```
   */
  static parseUrl(url: string): UrlComponents {
    const parsed = new URL(url);

    // Authority = host (includes port only if non-standard)
    // URL.host already includes port if non-standard, which matches RFC 9421
    const authority = parsed.host;

    // Path = pathname (e.g. "/api/v1/tasks")
    const path = parsed.pathname || "/";

    // Origin = scheme + host
    const origin = parsed.origin;

    return { authority, path, origin };
  }

  /**
   * Sign a message using EIP-191 personal_sign.
   *
   * Applies the prefix: `\x19Ethereum Signed Message:\n` + length + message
   * then signs with secp256k1 via viem.
   *
   * @param message - The signature base string to sign
   * @returns 65-byte signature (r[32] + s[32] + v[1])
   */
  private async eip191Sign(message: string): Promise<Uint8Array> {
    // viem's signMessage applies EIP-191 prefix internally
    const signature = await this.account.signMessage({ message });

    // Convert hex signature to bytes (remove "0x" prefix)
    return hexToBytes(signature);
  }
}

// ---------------------------------------------------------------------------
// Signature Base Construction (RFC 9421)
// ---------------------------------------------------------------------------

/**
 * Build the RFC 9421 signature base string.
 *
 * Format (lines joined with "\n", no trailing newline):
 * ```
 * "@method": POST
 * "@authority": api.execution.market
 * "@path": /api/v1/tasks
 * "content-digest": sha-256=:X48E...=:
 * "@signature-params": ("@method" "@authority" "@path" "content-digest");created=...;keyid="..."
 * ```
 *
 * Must exactly match the verifier's `_build_signature_base()` output.
 */
function buildSignatureBase(
  coveredComponents: string[],
  componentValues: Record<string, string>,
  sigParams: string,
): string {
  const lines: string[] = [];

  for (const component of coveredComponents) {
    const value = componentValues[component] ?? "";
    lines.push(`"${component}": ${value}`);
  }

  lines.push(`"@signature-params": ${sigParams}`);

  return lines.join("\n");
}

/**
 * Build the @signature-params value for the Signature-Input header.
 *
 * Matches the verifier's `_build_signature_params()`:
 *   - Components in parentheses: ("@method" "@authority" ...)
 *   - Parameters in fixed order: created, expires, nonce, keyid
 *   - Integer values are bare, string values are quoted
 *   - Joined with semicolons
 *
 * @example
 * ```
 * ("@method" "@authority" "@path" "content-digest");created=1700000000;expires=1700000300;nonce="abc123";keyid="erc8128:8453:0x857f..."
 * ```
 */
function buildSignatureParams(
  coveredComponents: string[],
  params: { created: number; expires: number; nonce: string; keyid: string },
): string {
  const compStr = coveredComponents.map((c) => `"${c}"`).join(" ");
  const parts = [`(${compStr})`];

  // Fixed order matching the verifier: created, expires, nonce, keyid
  parts.push(`created=${params.created}`);
  parts.push(`expires=${params.expires}`);
  parts.push(`nonce="${params.nonce}"`);
  parts.push(`keyid="${params.keyid}"`);

  return parts.join(";");
}

// ---------------------------------------------------------------------------
// Content-Digest (RFC 9530)
// ---------------------------------------------------------------------------

/**
 * Compute the Content-Digest header value for a request body.
 *
 * Format: `sha-256=:base64(sha256(body)):`
 *
 * @param body - The request body as a UTF-8 string
 * @returns Content-Digest header value
 */
function computeContentDigest(body: string): string {
  const hash = createHash("sha256").update(body, "utf8").digest("base64");
  return `sha-256=:${hash}:`;
}

// ---------------------------------------------------------------------------
// Hex / Byte utilities
// ---------------------------------------------------------------------------

/**
 * Convert a hex string (with "0x" prefix) to a Uint8Array.
 *
 * @param hex - Hex string like "0xabcdef..."
 * @returns Uint8Array of raw bytes
 */
function hexToBytes(hex: string): Uint8Array {
  const clean = hex.startsWith("0x") ? hex.slice(2) : hex;
  const bytes = new Uint8Array(clean.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(clean.substring(i * 2, i * 2 + 2), 16);
  }
  return bytes;
}

// ---------------------------------------------------------------------------
// Signed Fetch Wrapper
// ---------------------------------------------------------------------------

/** Options for the signed fetch helper (extends standard RequestInit) */
interface SignedFetchOptions extends RequestInit {
  /** Override the body for signing (if body is a non-string type) */
  bodyString?: string;
}

/**
 * Create a fetch wrapper that automatically signs requests with ERC-8128 headers.
 *
 * The returned function has the same signature as `fetch()` but injects
 * Content-Digest, Signature, and Signature-Input headers into every request.
 *
 * @param privateKey - Hex-encoded private key
 * @param chainId    - EVM chain ID (default: 8453 = Base)
 * @returns An async function with the same interface as `fetch()`
 *
 * @example
 * ```ts
 * const signedFetch = createSignedFetch("0xprivatekey", 8453);
 *
 * // POST with body — 3 ERC-8128 headers added
 * const res = await signedFetch("https://api.execution.market/api/v1/tasks", {
 *   method: "POST",
 *   headers: { "Content-Type": "application/json" },
 *   body: JSON.stringify({ title: "Buy coffee", bounty_usdc: "0.10" }),
 * });
 *
 * // GET — 2 ERC-8128 headers added (no content-digest)
 * const tasks = await signedFetch("https://api.execution.market/api/v1/tasks");
 * ```
 */
export function createSignedFetch(
  privateKey: string,
  chainId: number = 8453,
): (url: string, init?: SignedFetchOptions) => Promise<Response> {
  const signer = new EIP8128Signer(privateKey, chainId);

  return async (url: string, init?: SignedFetchOptions): Promise<Response> => {
    const method = (init?.method ?? "GET").toUpperCase();

    // Resolve the body string for signing
    let bodyString: string | undefined = init?.bodyString;
    if (bodyString == null && init?.body != null) {
      if (typeof init.body === "string") {
        bodyString = init.body;
      } else {
        // For non-string bodies (ArrayBuffer, ReadableStream, etc.),
        // caller should provide bodyString explicitly
        throw new Error(
          "EIP-8128 signed fetch requires a string body. " +
            "Pass the body as a string or use the bodyString option.",
        );
      }
    }

    // Sign the request
    const sigHeaders = await signer.signRequest(method, url, bodyString);

    // Merge signature headers with any existing headers
    const existingHeaders = new Headers(init?.headers as HeadersInit | undefined);
    for (const [key, value] of Object.entries(sigHeaders)) {
      existingHeaders.set(key, value);
    }

    // Execute the fetch with merged headers
    return fetch(url, {
      ...init,
      method,
      headers: existingHeaders,
    });
  };
}
